lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -1097,15 +1143,38 @@ estimateNumberOfWitnesses utxo txbody@(Cardano.TxBody txbodycontent) =
         length txExtraKeyWits' +
         length txWithdrawals' +
         txUpdateProposal' +
-        txCerts +
+        fromIntegral txCerts +
         scriptVkWitsUpperBound
   where
     (Cardano.ShelleyTxBody _ _ scripts _ _ _) = txbody
 
     dummyKeyRole = Payment
 
-    isStakeKeyRegCert (Cardano.StakeAddressRegistrationCertificate _) = True
-    isStakeKeyRegCert _ = False
+    apiScriptHashes =
+        let walletScripts = mapMaybe toTimelockScript scripts
+            toScriptHash =
+                Cardano.hashScript
+                . Cardano.SimpleScript Cardano.SimpleScriptV2
+                . toCardanoSimpleScript
+            toPair s = (s, toScriptHash s)
+        in toPair <$> walletScripts
+
+    estimateWitNumForCred = \case
+        Cardano.StakeCredentialByKey _ -> 1
+        Cardano.StakeCredentialByScript scriptHash ->
+            let pair = filter (\(_,sh) -> sh == scriptHash) apiScriptHashes
+            in case pair of
+                (delScript,_):[] ->
+                    estimateMaxWitnessRequiredPerInput delScript
+                _ -> error "there should be delegation script in the tx body"
+
+    estimateDelegSigningKeys = \case
+        Cardano.StakeAddressRegistrationCertificate _ -> 0
+        Cardano.StakeAddressDeregistrationCertificate cred ->
+            estimateWitNumForCred cred
+        Cardano.StakeAddressDelegationCertificate cred _ ->
+            estimateWitNumForCred cred
```

Comment:
I don't think we need to resolve the scripts from ByScript credentials. We already have `scriptVkWitsUpperBound`, and could handle it in a simpler and possible safer way by:
1. Make `estimateWitNumForCred` only return `1` for key credentials and 0 otherwise.
2. Drop `apiScriptHashes` - we don't need the resolution

I tried this locally and it fails with `(FeeTooSmallUTxO (Coin 176900) (Coin 166800)`. But I suspect this is because we somehow sign with one witness too much, probably because of duplication between:

```
        , mapMaybe mkWdrlCertWitness certs
        , mapMaybe mkStakingScriptWitness stakingScriptsKeyHashes
```